### PR TITLE
[FIX] web: increase file upload to 128Mb

### PR DESCRIPTION
**Pull request by william-andre**
This is the limit of the SAAS servers.

This is done while trying to improve the experience of Winbooks import for larger databases.
[opw-2251443](https://www.odoo.com/web#active_id=2251443&cids=1&id=2251443&model=project.task)
Enterprise PR: https://github.com/odoo/enterprise/pull/14370

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by aab-odoo** [comment]
@william-andre We increased the limit to 64gb in saas-13.3. Did you know? Maybe it is enough..

**Comment by william-andre** [comment]
@aab-odoo 64Mb, right?
128Mb is the limit set on the servers on the SAAS. We saw that you made the change in 14.0, but we couldn't see why you chose that value which seems arbitrary.
Since the servers are configured to go up to 128Mb, why have a lower limit in javascript?

**Comment by aab-odoo** [comment]
I honestly don't know. In the [commit](https://github.com/odoo/odoo/commit/258a04664ad141cf6ce125a832aaba263c0aa9cd) message, Géry mentions a similar limit for nginx.

**Comment by william-andre** [comment]
Not everyone uses nginx, and even if you do, this is configurable. We are using 128Mb on our servers so it makes sense to have the hardcore value to this...

**Review by yezz123 — COMMENTED** [review]
why not adding 256mb for more ?

**Review by odony — COMMENTED** [review]
tl;dr: Looks ok, unless we want to go one step further and make the client-side limit configurable - which would be even better.

Long version: There are several things at play here.

1. In order to protect servers from trivial attacks caused by large files, a key step for secure deployments is to configure a request size limit on the web server that is proxying requests (nginx, apache, etc.). That limit is applied *server-side* to all requests including file uploads. It may be set to any arbitrary value depending on the requirements for a particular deployment. Note: the [default for nginx](http://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size) (our recommended web server) is in fact 1MB.

2. On our cloud service, as of 2020, the above-mentioned server-side limit is generally 64 or 128MB, in order to allow for largish regular file uploads yet block videos and bigger files. It may change from time to time as a trade-off between business needs and security policies, as technologies evolve. 

3. The limit applied in the web client (= this PR) is not a security feature (as all things client-side) but a usability feature. Instead of allowing large uploads that may trigger the server-side limit and result in an obscure error message, the web client blocks large files with a proper error, also saving some bandwidth and time. Ideally this limit should be configurable on each deployment (e.g. as a system parameter), so that administrators can match it with the server-side limit on their deployment.

**Comment by william-andre** [comment]
@odony I went for a `ir.config_parameter`, not sure if you expected a server parameter instead?

File: addons/web/static/src/js/core/utils.js [hunk 1]
```diff
@@ -354,7 +354,7 @@ var utils = {
      * @param {Number} size number of bytes
      */
     human_size: function (size) {
-        var units = _t("Bytes|Kb|Mb|Gb|Tb|Pb|Eb|Zb|Yb").split('|');
```
Comment (odony) [inline addons/web/static/src/js/core/utils.js:357]: I know it's tempting to be "more correct", but using the technical IEC units will just confuse users, as they've never heard of those. Users don't count bytes manually and don't care about "more correct" units. The only reference users have about file size is what [their OS tells them](https://en.wikipedia.org/wiki/Binary_prefix#Operating_systems). Plus this number is only useful as a rough estimation anyway. So I would stick with the commonly accepted units of "kB" or "KB" for 1024 bytes as an acceptable trade-off for most users, because that's what Windows displays.

But most importantly, I would not change that in stable at all, as it's not strictly necessary, and our end-users are used to it. Feel free to make a separate PR for changing "Kb" into "KB" in master ;-)

PS: https://ux.stackexchange.com/a/13850 and https://blog.codinghorror.com/gigabyte-decimal-vs-binary/

File: addons/web/models/ir_http.py [hunk 2]
```diff
@@ -41,6 +41,7 @@ def session_info(self):
             "company_id": user.company_id.id if request.session.uid else None,  # YTI TODO: Remove this from the user context
             "partner_id": user.partner_id.id if request.session.uid and user.partner_id else None,
             "web.base.url": self.env['ir.config_parameter'].sudo().get_param('web.base.url', default=''),
+            "max_file_size": int(self.env['ir.config_parameter'].sudo().get_param('web.max_file_size', default=128 * 1024 * 1024)),  # 128MiB
```
Comment (odony) [inline addons/web/models/ir_http.py:44]: Could we make the key slightly more specific? (also computing the default size on a different line could make the line shorter and the code more readable)

```suggestion
            "max_file_size": int(self.env['ir.config_parameter'].sudo().get_param('web.max_file_upload_size', default=128 * 1024 * 1024)),  # 128MiB
```

File: addons/web/static/src/js/fields/basic_fields.js [hunk 3]
```diff
@@ -1663,7 +1663,7 @@ var AbstractFieldBinary = AbstractField.extend({
         this._super.apply(this, arguments);
         this.fields = record.fields;
         this.useFileAPI = !!window.FileReader;
-        this.max_upload_size = 25 * 1024 * 1024; // 25Mo
+        this.max_upload_size = odoo.session_info.max_file_size;
```
Comment (odony) [inline addons/web/static/src/js/fields/basic_fields.js:1666]: I'd rather keep a default here, in case the new JS code runs before the Python code gets reloaded (safer for a stable version).

```suggestion
        this.max_upload_size = odoo.session_info.max_file_size || 128 * 1024 * 1024;
```

**Review by odony — COMMENTED** [review]
Yes, I meant an `ir_config.parameter`, thanks! :+1:

File: addons/web/static/src/js/core/utils.js (hunk 1 above)
Comment (william-andre) [reply addons/web/static/src/js/core/utils.js]: We are letting hard drive manufacturer and the ISPs lie to users if everybody uses the same notation for different things in the same context; but ok, I get your point.

**Review by yezz123 — APPROVED** [review]
As a suggestion!
make the upload up to 256mb!

File: addons/web/static/tests/fields/basic_fields_tests.js [hunk 4]
```diff
@@ -19,6 +19,10 @@ var DebouncedField = basicFields.DebouncedField;
 var JournalDashboardGraph = basicFields.JournalDashboardGraph;
 var _t = core._t;
```
Comment (antonylesuisse) [inline addons/web/static/tests/fields/basic_fields_tests.js:21]: not needed ? as we have a fall back in basic_fields (odoo.session_info.max_file_upload_size || 128 * 1024 * 1024);

Comment (william-andre) [reply addons/web/static/tests/fields/basic_fields_tests.js:21]: `session_info` is undefined in the test

File: addons/web/static/src/js/fields/basic_fields.js [hunk 5]
```diff
@@ -1663,7 +1663,7 @@ var AbstractFieldBinary = AbstractField.extend({
         this._super.apply(this, arguments);
         this.fields = record.fields;
         this.useFileAPI = !!window.FileReader;
-        this.max_upload_size = 25 * 1024 * 1024; // 25Mo
+        this.max_upload_size = odoo.session_info.max_file_upload_size || 128 * 1024 * 1024;
```
Comment (aab-odoo) [inline addons/web/static/src/js/fields/basic_fields.js:1666]:
```suggestion
        this.max_upload_size = session.max_file_upload_size || 128 * 1024 * 1024;
```
That way, you should no longer need to patch anything in the tests

**Comment by william-andre** [comment]
@aab-odoo Can we merge this please? I have yet another issue with the size of a file.

**Review by aab-odoo — APPROVED** [review]

**Comment by aab-odoo** [comment]
robodoo r+
